### PR TITLE
fix some inspection errors & warnings

### PR DESCRIPTION
**Pull request by 1fish2**
(Please check. Some of these edits are just guesses. pytest is happy.)

* Fix some PyCharm inspection issues, including some PEP8 style points but I turned off several inspection types. Those settings file aren't checked in since `.idea/` is .gitignored but I could still add the inspection profile if you want.
* Drop some Python 2 compatibility, which means for example we can use `os.makedirs(p, exist_ok=True)`.
* Make `toy_control` put output into `out/` so we needn't .gitignore it or explain it.
* `class Datum` could use unit tests and more docs to explain the weird things it does and how to use it. I didn't add `super().__init__()` yet for lack of tests, but a quick experiment went OK.
* `# noinspection PyMissingConstructor` hides a warning about not calling `super().__init__()`.
* (There are lots of `{}` mutable default args here but no straightforward workaround in Python.)
* `make_network.py` has tab indentation (also `compartments.rst`). This PR could change it to spaces. It fixes the mixed indentation cases.

Inspections not fixed, ignored, or marked TODO:
* Duplicate code fragments.
* Packages not listed in project requirements. I could change `requirements.txt` for that.
* Some PEP 8 style deets.
* A type warning in experiment.py `test_multi()`.
* Unused imports in `setup.py` -- some of them might still matter.
* Unused locals.

### CODE REVIEW

**Review by eagmon — COMMENTED** [review]
These all look great, thanks @1fish2!  I only felt uneasy about setting ```self.state = []``` in ```vivarium/core/experiment.py```, line 1339.  

I would be okay with having inspection code via ```.idea/``` -- but would that find anything beyond the pylint approach that @U8NWXD started in #22? 

And instead of ```# noinspection PyMissingConstructor```, should we just add the constructor?

File: vivarium/core/composition.py [hunk 1]
```diff
@@ -299,7 +297,7 @@ def agent_environment_experiment(
     if settings.get('agent_names') is True:
         # add an AgentNames processes, which saves the current agent names
         # to as store at the top level of the hierarchy
-        processes['agent_names'] = AgentNames({})
+        # processes['agent_names'] = AgentNames({})  # TODO(jerry): vivarium_cell.processes.AgentNames?
```
[the remark this answers is not in the document]
Comment (eagmon) [reply vivarium/core/composition.py:300]: Actually, let's just leave this commented for now. I'll remember to uncomment it if I decide to bring ```AgentNames``` back in the future.

File: vivarium/core/experiment.py [hunk 2]
```diff
@@ -1336,6 +1336,7 @@ def __init__(self, config):
             self.processes,
             self.topology,
             self.initial_state)
+        self.states = []  # TODO(jerry): What value does self.update() need?
```
[the remark this answers is not in the document]
Comment (1fish2) [reply vivarium/core/experiment.py:1339]: Currently the `Experiment` class does not define or set the instance variable `self.states` but its `update()` method needs that to be set. What code sets it? Does it expect a subclass to set it? Even then, it should define it (and probably document expectations on subclasses) so it defaults properly and so dev tools like mypy will know what type to expect.

I'm just guessing that `update()` expects `self.states` to be a list. This initialization shouldn't replace any value unless something is sneaking in during or before the `__init__()` method somehow, like perhaps a subclass `__init__()` method that sets it before calling `super().__init__()`.

**Comment by 1fish2** [comment]
> I only felt uneasy about setting `self.state = []` in vivarium/core/experiment.py, line 1339.

Where's the code to set `self.state`?

> I would be okay with having inspection code via `.idea/` -- but would that find anything beyond the pylint approach that @U8NWXD started in #22?

pylint, PyCharm, mypy, and other static analysis tools check for different, but overlapping, issues. E.g. pylint doesn't say it includes a type checker. So it's helpful to run more than one tool. (It is possible to run PyCharm's inspections from the command line, if it's installed, but I used mypy to check types in wcEcoli CI because it's free and easy to install.) In any case the only effect of an inspection config file in `.idea/` would be to hide warnings in the PyCharm editor and from PyCharm "Inspect Code..." output, e.g. don't warn about:

* non-PEP8 names (one argument uses camelCase, some others have capital letters, maybe to denote matrices),
* other PEP8 style issues like too few and too many blank lines between top level definitions, or comments that start with `##`, or single-quoted docstrings including `'''...'''`,
* code that's not compatible with Python before 3.6.

> And instead of `# noinspection PyMissingConstructor`, should we just add the constructor?

Yes, `super().__init__()`. I didn't add it in some cases because I wasn't sure it was just an oversight. Experimenting shows that it's fine in a case like `class Datum(dict)`.

How does `Datum` get used? This code is a bit weird, subclassing `dict`, not callint `super().__init__()`, and setting `self.__dict__ = self`. Does it set `.__dict__` to enable writing, say, `datum.x` in place of `datum.['x']`? If that's the case, it'd be safer to instead define

    def __getattr__(self, item):
        return self[item]

This gets called after regular attribute lookup fails, so it shouldn't interfere with other mechanisms like properties. It probably saves memory by not adding an instance dict to each of these dicts. But it might be a bit slower by running a line of Python code in that case rather than doing the attribute lookup entirely in C code.

File: vivarium/core/experiment.py [hunk 3]
```diff
@@ -576,7 +576,7 @@ def reduce(self, reducer, initial=None):
     def reduce_to(self, path, reducer, initial=None):
         value = self.reduce(reducer, initial)
         assoc_path({}, path, value)
-        self.apply_update(update)
+        self.apply_update({})  # TODO(jerry): ? This was `apply_update(update)` but update is undefined
```
[the remark this answers is not in the document]
Comment (1fish2) [reply vivarium/core/experiment.py:579]: Please.

[the remark this answers is not in the document]
Comment (eagmon) [reply vivarium/core/experiment.py:579]: This was not even being used, I'm removing it.

File: vivarium/core/experiment.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eagmon) [reply vivarium/core/experiment.py:1339]: Oh I see, this is ```self.states```, not ```self.state```. This is mysterious...

File: vivarium/core/process.py [hunk 4]
```diff
@@ -232,24 +231,11 @@ class Process(Generator):
...
-        self.schema_override = {}
-        if '_schema' in parameters:
-            self.schema_override = parameters.pop('_schema')
-
-        self.parallel = False
-        if '_parallel' in parameters:
-            self.parallel = parameters.pop('_parallel')
-
-        deep_merge(self.parameters, parameters)
+        self.parallel = self.config.pop('_parallel', False)
```
Comment (eagmon) [inline vivarium/core/process.py:238]: This is a wonderful simplification! thank you!

**Review by eagmon — APPROVED** [review]
This looks great! Thank you.  The ```reduce_to``` in ```experiment.py``` was not being used, so I removed it.  I added AgentNames process back in -- it had found its way to ```vivarium-cell```, but it's more fitting here. I am still unsure what ```self.states``` is doing  in ```Experiment```, but this is good to merge for now -- I'll keep that issue in mind.

File: vivarium/core/experiment.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eagmon) [reply vivarium/core/experiment.py:1333]: @1fish2 -- I'm pretty sure the ```self.states``` in ```self.update()``` was a typo -- it was only being used if verbose.  I am fixing the typo, and removing this line.

**Comment by eagmon** [comment]
@1fish2 -- I just went through this again, and think it's ready to merge. There are a couple TODOs left in there that look pretty simple -- come back to them later, or fix them before the merge?

**Comment by 1fish2** [comment]
Great! We can leave the TODOs for later. I just remembered about adding the .idea inspection profile. I'll do that and double check.

* run_plots() in control.py refers to undefined `plot_id`. I'll change it to `run_plots`.
* make_network.py uses TABs. A PyCharm command changes it to spaces.
* In Python 3, `class Foo:` is the same as `class Foo(object):`. I think a style inspection warned about the latter, but it's fine that way.
* In `Process.__init__()`, the previous code replaced the base class self.config with an empty `{}`. Is it OK now to keep the inherited one, including `Generator.defaults`? The latest change certainly makes it simpler and leaves the class less tightly coupled to its base class.
* In `Experiment.update()`, `for state_id in self.state:` requires `self.state` to be `Iterable` but the type checker points out it's a `Store`, which does not look `Iterable`. Should it have an `__iter__()` method? Or should this code fetch a part of it?

**Comment by 1fish2** [comment]
* `self.state[state_id]` also won't work if `self.state` is a `Store`.

Some type check warnings that we can leave for later:
* In `process_in_experiment()`, a list doesn't have a `get()` method in  `timeline.get('paths', {})`. Should `timeline` be a dict, in which case `timeline = settings.get('timeline', [])` should default to `{}`?
* In `test_multi()`, the type checker doesn't like `Experiment(dict(proton, invoke=multi.invoke))`.  
  `Unexpected type(s): (Dict[str, Union[Dict[str, Union[Proton, Dict[str, Dict[str, Electron]]]], Dict[str, Union[Dict[str, Union[Tuple[str, str], ...]], Dict[str, ...]]], Dict[str, Union[..., ...]]]], (process: Any, interval: Any, states: Any) -> Any) Possible types: (Mapping[str, Union[Dict[str, Union[Proton, Dict[str, Dict[str, Electron]]]], Dict[str, Union[Dict[str, Union[Tuple[str, str], ...]], Dict[str, ...]]], Dict[str, Union[..., ...]]]], Union[Dict[str, Union[Proton, Dict[str, Dict[str, Electron]]]], Dict[str, Union[Dict[str, Union[Tuple[str, str], Dict[str, ...]]], Dict[str, Dict[str, ...]]]], Dict[str, Union[Dict[str, float], Dict[str, ...]]]]) (Iterable[Tuple[Any, Any]], _VT)`.

**Review by 1fish2 — COMMENTED** [review]
[Did I forget some pending comments on GitHub? It says I started a review. I'll submit this to see what happens.]

**Comment by 1fish2** [comment]
I think this is good to go. Do you want to recheck, @eagmon ? Of course we can always add more unit tests to ensure I didn't break anything.

It'd be good to distinguish private from public library definitions via `_` names or `all` lists so we know which definitions can be changed or removed without impacting clients.

**Review by eagmon — APPROVED** [review]
This looks excellent, thanks @1fish2! Go ahead and merge it in. I'll make a release of vivarium-core sometime soon and test this in all of our other repos.

**Pull request merged**